File: src/script-py/0004-median-of-two-sorted-arrays.py
# File Name: 0004-MEDIAN-OF-TWO-SORTED-ARRAYS.PY
# Author: Qiangua
# Email: caicailiushui@protonmail.com
# Copyright © 2021, Qiangua, all rights reserved. 
# Created:  7 November 2021 
#  -*- coding: utf-8 -*-

from typing import List
class Solution:
    def findMedianSortedArrays(self,nums1: List[int],nums2: List[int]) -> float:
        l1,l2 = len(nums1),len(nums2)
        start,end = 0,l1
        mid_posi = (l1 + l2 + 1)//2
        inf = 2**40
        if l1 > l2:
            return self.findMedianSortedArrays(nums2,nums1)
        mid_posi:int = (l1 + l2 + 1)//2
        while start <= end: 
            mid:int = (start + end)//2
            left1size:int = mid
            left2size:int = mid_posi - mid
            # print(mid,left1size,left2size)
            left1ele = nums1[left1size - 1] if left1size > 0 else -inf
            left2ele = nums2[left2size - 1] if left2size > 0 else -inf
            right1ele = nums1[left1size] if left1size < l1 else inf
            right2ele = nums2[left2size] if left2size < l2 else inf
            if left1ele <= right2ele and left2ele <= right1ele:
               if (l1 + l2) % 2 == 0:
                   return (max(left1ele,left2ele) + min(right1ele,right2ele))/2.0
               else:
                   return max(left1ele,left2ele)
            elif left1ele > right2ele:
                end = mid - 1
            else:
                start = mid + 1
    
 



    
import unittest                
class TestLeetcode(unittest.TestCase):
    def test1(self):
        n1 = [1,2,3]
        n2 = [3]
        # n2 = [4,5,6]
        sol = Solution()
        actual = sol.findMedianSortedArrays(n1,n2)
        # expect = 3.5
        expect = 2.5
        self.assertEqual(expect,actual)

        
if __name__ == "__main__":
    unittest.main() 

File: src/script-cxx/0004-median-of-two-sorted-arrays.cpp
// File Name: 0004-MEDIAN-OF-TWO-SORTED-ARRAYS.CPP
// Author: Qiangua
// Email: caicailiushui@protonmail.com
// Copyright © 2021, Qiangua, all rights reserved. 
// Created:  6 November 2021 

#include "headers.h"
#include <gtest/gtest.h>

class Solution{
public: 
    double findMedianSortedArrays(vector<int>& A, vector<int>& B)
        {
            int n = A.size(), m = B.size();
            if (n > m) {
                return findMedianSortedArrays(B, A); // Swapping to make A smaller
            }
 
            int start = 0;
            int end = n;
            int mid_posi = (n + m + 1) / 2;
 
            while (start <= end) {
                int mid = (start + end) / 2;
                int leftAsize = mid;
                int leftBsize = mid_posi - mid;
                int leftA = (leftAsize > 0) ? A[leftAsize - 1] : INT_MIN; // checking overflow of indices
                int leftB = (leftBsize > 0) ? B[leftBsize - 1] : INT_MIN;
                int rightA = (leftAsize < n) ? A[leftAsize] : INT_MAX;
                int rightB = (leftBsize < m) ? B[leftBsize] : INT_MAX;
                // if correct partition is done
                if (leftA <= rightB and leftB <= rightA) {
                    // switch ( (m + n) % 2) {
                    // case 0:
                    //     return (max(leftA, leftB) + min(rightA, rightB)) / 2.0;
                    // case 1:
                    //     return max(leftA, leftB);
                    // }
                    if ((m + n) % 2 == 0){
                        return (max(leftA, leftB) + min(rightA, rightB)) / 2.0;
                    }
                    else {
                        return max(leftA, leftB);
                    }
                }
                else if (leftA > rightB) {
                    end = mid - 1;
                }
                else
                start = mid + 1;
            }
            return 0.0;
        }
};

TEST(TTT, TTT) {
    Solution s;
    // vector<int> nums1 = { 1,2,3 },nums2 = { 4,5,6 };
    vector<int> nums1 = { 1,2,3 },nums2 = { 4,5,6 };
    double expect = 3.5;
    double actual;
    actual = s.findMedianSortedArrays(nums1,nums2);
    EXPECT_EQ(expect,actual);
}


int main(int argc, char **argv) {
    cout << "Problem: 0004-MEDIAN-OF-TWO-SORTED-ARRAYS.CPP" << endl;
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
